source/grid/grid_generator.cc

```diff
@@ -5485,48 +5544,106 @@ namespace GridGenerator
+            // children
+            if (temporary_map_boundaryCell_face[i1]
+                  .second.first->has_children())
+              {
+                Assert(temporary_map_boundaryCell_face[i1]
+                           .second.first->refinement_case() ==
+                         RefinementCase<dim - 1>::isotropic_refinement,
+                       ExcNotImplemented()); // algorithm only works for
+                                             // isotropic refinement!
+                temporary_map_boundaryCell_face[i1].first->set_refine_flag();
+                cells_refined.push_back(i1);
+                changed = true;
+              }
+          }
 
+        // if cells have been marked for refinement (i.e., presently deepest
+        // level is not the deepest level of the volume mesh)
         if (changed)
           {
+            // do actual refinement
             const_cast<Triangulation<dim - 1, spacedim> &>(
               surface_mesh.get_triangulation())
               .execute_coarsening_and_refinement();
 
-            for (typename MeshType<dim - 1, spacedim>::cell_iterator
-                   surface_cell = surface_mesh.begin();
-                 surface_cell != surface_mesh.end();
-                 ++surface_cell)
-              for (unsigned int c = 0; c < surface_cell->n_children(); c++)
-                if (surface_to_volume_mapping.find(surface_cell->child(c)) ==
-                    surface_to_volume_mapping.end())
-                  surface_to_volume_mapping[surface_cell->child(c)] =
-                    surface_to_volume_mapping[surface_cell]->child(c);
+            // add new level of cells to temporaryMap_boundaryCell_face
+            index_cells_deepest_level = temporary_map_boundaryCell_face.size();
+            for (unsigned int i2 = 0; i2 < cells_refined.size(); i2++)
+              {
+                typename MeshType<dim - 1, spacedim>::cell_iterator
+                  refined_cell =
+                    temporary_map_boundaryCell_face[cells_refined[i2]].first;
+                typename MeshType<dim, spacedim>::face_iterator refined_face =
+                  temporary_map_boundaryCell_face[cells_refined[i2]]
+                    .second.first;
+                unsigned int refined_face_number =
+                  temporary_map_boundaryCell_face[cells_refined[i2]]
+                    .second.second;
+                for (unsigned int i3 = 0; i3 < refined_cell->n_children(); i3++)
+                  // at this point, the swapping of vertices done earlier must
+                  // be taken into account to get the right association between
+                  // volume faces and boundary cells!
+                  temporary_map_boundaryCell_face.push_back(std::make_pair(
+                    refined_cell->child(swap_matrix[refined_face_number][i3]),
+                    std::make_pair(refined_face->child(i3),
+                                   refined_face_number)));
+              }
```

Comment:
This would be easier to understand if we used more specific array indices than `i1`, `i2`, and `i3`. How about `cell_n`, `refined_cell_n`, and `child_n`?

Reply:
I don't have a strong preference, but good loop variables would make this easier to read.

Reply:
Agreed, although I don think that it helps much at this point (really descriptive names would have to be prohibitively long here)